File: neetcode/20. Valid Parentheses/20.java
class Solution {
    public boolean isValid(String s) {
        Deque<Character> stack = new LinkedList<>();

        for (char c : s.toCharArray()) {
            if (c == '(' || c == '{' || c == '[') {
                stack.push(c);
            } else {
                if (stack.isEmpty()) return false;
                else {
                    if ((c == ')' && stack.peek() == '(') ||
                            (c == '}' && stack.peek() == '{') ||
                            (c == ']' && stack.peek() == '['))
                    {
                        stack.pop();
                    }
                    else return false;
                }
            }
        }

        return stack.isEmpty();
    }
}

File: neetcode/20. Valid Parentheses/20.kt
class Solution {
    fun isValid(s: String): Boolean {
        val arr = CharArray(s.length)

        var top = -1

        for (c in s.toCharArray()) {
            if (c == '(' || c == '{' || c == '[') {
                arr[++top] = c
            } else {
                if (top == -1) return false

                if ((c == ')' && arr[top] == '(') ||
                    (c == '}' && arr[top] == '{') ||
                    (c == ']' && arr[top] == '[')) {
                    top--
                }
                else return false
            }
        }

        return top == -1
    }
}
